meetings/working-groups/field-keyword/FK-2024-05-26.md

```diff
@@ -0,0 +1,255 @@
+
+    while (true)
+    {
+        if (rand())
+            Prop_get();
+        else
+            Prop_set(...non-null-value...)
+    }
+
+    T Prop_get() { ...getter body... }
+    void Prop_set(T value) { ...setter body... }
+}
+```
+
+This looks a bit daunting at first, but comes with a simple general intuition:
+
+1. The language/compiler already support precise NRT flow analysis of locals and local functions within a method.
+1. The backing field is rethought of a just a nullable local field.
+1. If initialized at the property declaration site, there is an up front assignment to that backing field solely for the purpose of updating the initial flow state.
+1. The normal constructor statements are analyzed, with any calls to the auto-props in the type now being calls to the local functions corrsponding to the accessors.  This ensures precise NRT analysis at every call site.
+1. After the constructor statements finish, the property is analyzed again in a fashion that simulates how the getter/setter can be called randomly in any order after the constructor concludes.
+
+With this formalization, precise NRT analysis just "falls out". For example, if the user had:
+
+```c#
+string Prop { get => field ?? ""; } // or
+string Prop { get => field ??= ComputeInitialValue(); }
+```
+
+then they would have to do nothing else (no need for any updating of the constructor).  Similarly, the following would be known to be safe without warnings:
+
+```c#
+string Prop { get => field; set => field = value.Trim(); }
+
+public C()
+{
+    this.Prop = "";
+}
+```
+
+However, this would properly report a warning:
+
+```c#
+// warning that `field` in the getter might be null.
+string Prop { get => field; set => field = value is "" ? null : value.Trim(); }
+
+public C()
+{
+    this.Prop = "";
+}
+```
+
+Everything is precise here because we are performing real NRT analysis across the constructor statements, and for after the constructor finishes up.
+
+The downside here though is complexity of implementation and potential computational costs having to perform this analysis for each constructor in the type.  It would also be critical for diagnostic messages to clearly state both where a null-ref issue might exist, and from what constructor or post-constructor path was potentially causing it.
+
+## Approach C. Backing field's type may or may not be the nullable version of the property's type.
+
+The final approach we are exploring is one where the backing field is *conditionally* nullable.  The intuition here is to *first* do an initial analysis pass of the get-accessor starting with the assumption that the backing field *is* nullable.  If the conclusion of that analysis is that all exit points of the accessor leave the field in the non-null state, then the backing field will be considered nullable.  Otherwise, it will be considered non-nullable, and the existing language rules will take effect.
```

Comment:
just a random observation. We would have to be a little tricky in the compiler if pursuing this solution. I think the nullable annotation of the field symbol can't depend on the results of a flow analysis. We would have to put this info on the side somewhere in the TypeCompilationState or something like that. Luckily we would only be doing a propagation of this info to other members of the same type.

The flow dependency we would finally be creating here is: in order to nullable analyze any constructor, we first need to search the type members for getters which use `field`, then bind and analyze all such getters with field in the maybe-null initial state. But I don't think it would cycle around or propagate beyond that.

Reply:
> just a random observation. We would have to be a little tricky in the compiler if pursuing this solution. I think the nullable annotation of the field symbol can't depend on the results of a flow analysis

*Today* :)  

> But I don't think it would cycle around or propagate beyond that.

Yup.  It seems to be ordered ok.  Especially as a property can only see into its own `field`, and can just trust the signature of any other property in the type.  So there's a pass *prior* to constructors, taht figures this out for all the auto-props.  Then we do the constructor pass.

>  I think the nullable annotation of the field symbol can't depend on the results of a flow analysis. 

fair.  i'm not trying to be precise in terms of compiler logic/storage, just in the *general intuition* of how this would work.  Basically "do an initial analysis pass of hte getter to see if null can flow out or not.  if not, then the field is allowed to be nullable.  Otherwise it is not allowed to be".  The actual mechanics are totally up to you :)